File: mobile/src/census.ts
/**
 * 여러 서버의 응답을 한 화면으로 합치는 순수 로직. Tauri/DOM 없이 테스트된다.
 *
 * 여기서 지키는 규칙 하나: **대답하지 않은 서버는 조용히 사라지지 않는다.**
 * 세션이 목록에 없으면 사용자는 그 세션이 죽었다고 읽는다. 그래서 합친 목록과
 * 나란히, 대답하지 못한 서버가 각각 무엇 때문이었는지 항상 함께 내놓는다.
 */

import type { RemoteSession } from "./sessions";
import { sessionTitle } from "./sessions";

/** Rust `census::ProbeOutcome`와 같은 모양. */
export type ProbeOutcome =
  | { state: "listed"; sessions: RemoteSession[] }
  | { state: "not_provisioned"; detail: string }
  | { state: "not_configured"; code: string; detail: string }
  | { state: "unreachable"; code: string; detail: string }
  | { state: "timed_out"; seconds: number }
  | { state: "not_attempted" };

/** Rust `census::ServerReport`. 서버 하나당 정확히 하나. */
export interface ServerReport {
  server_id: string;
  server_label: string;
  outcome: ProbeOutcome;
}

/** 합친 목록의 한 줄. 어느 서버의 세션인지 항상 달고 다닌다. */
export interface CensusRow {
  serverId: string;
  serverLabel: string;
  /**
   * Whether the server answered *this* census.
   *
   * False rows come from [`rememberListings`]: the server listed them once and
   * did not answer now. They are drawn, dimmed and unopenable, rather than
   * dropped — for the reason at the top of this file.
   */
  reachable: boolean;
  session: RemoteSession;
}

/** What one server last said, kept for the census in which it stays silent. */
export interface RememberedListing {
  serverLabel: string;
  sessions: RemoteSession[];
}

/**
 * Carry each server's last good listing across a census it failed.
 *
 * The paired-computer path already does exactly this (`app.ts` keeps the
 * sessions and flips `reachable`), and the two transports have to agree: a
 * project is unreachable, not gone, whichever way the phone reaches it.
 * Without this the SSH half of a desktop empties the moment a server times
 * out, and the screen is left with no rows to dim and only a count to
 * apologise with.
 *
 * A server that has never answered contributes nothing: there is no listing to
 * remember, and inventing rows for it would be worse than saying nothing.
 */
export function rememberListings(
  previous: Readonly<Record<string, RememberedListing>>,
  reports: readonly ServerReport[],
): Record<string, RememberedListing> {
  const next = { ...previous };
  for (const report of reports) {
    if (report.outcome.state !== "listed") continue;
    next[report.server_id] = {
      serverLabel: report.server_label,
      sessions: report.outcome.sessions,
    };
  }
  return next;
}

/** 대답하지 못한 서버 한 대와, 사용자가 무엇을 고쳐야 하는지. */
export interface CensusFailure {
  serverId: string;
  serverLabel: string;
  /** 화면에 그대로 쓰는 한국어 문장. */
  message: string;
  /** 세부. 서버가 준 말이라 번역하지 않는다. */
  detail?: string;
}

/**
 * 서버가 대답하지 못한 이유를 사용자가 할 일로 바꾼다.
 *
 * hmux가 없는 서버는 SSH가 완전히 정상인 채로 셸 오류를 돌려준다. 그것을
 * 프로토콜 실패라고 부르면 사용자는 네트워크를 들여다보게 된다 — 고칠 곳은
 * 서버의 PATH다.
 */
export function describeFailure(report: ServerReport): CensusFailure | undefined {
  const base = { serverId: report.server_id, serverLabel: report.server_label };
  switch (report.outcome.state) {
    case "listed":
      return undefined;
    case "not_provisioned":
      return { ...base, message: "hmux가 설치되어 있지 않습니다", detail: report.outcome.detail };
    case "not_configured":
      return {
        ...base,
        message: "이 기기에 연결 정보가 없습니다",
        detail: report.outcome.detail,
      };
    case "unreachable":
      return { ...base, message: "연결하지 못했습니다", detail: report.outcome.detail };
    case "timed_out":
      return {
        ...base,
        message: `${report.outcome.seconds}초 안에 응답을 마치지 못했습니다`,
      };
    case "not_attempted":
      return { ...base, message: "시간이 모자라 물어보지 못했습니다" };
  }
}

/**
 * 모든 서버의 세션을 한 목록으로.
 *
 * 정렬은 붙을 수 있는 세션 먼저, 그다음 서버 이름, 그다음 세션 이름. 서버별로
 * 묶지 않는 이유: 사용자가 찾는 것은 "어느 서버"가 아니라 "어제 남겨둔 그
 * 에이전트"이고, 서버는 그것을 찾은 다음에 필요한 정보다.
 */
export function mergeSessions(
  reports: readonly ServerReport[],
  /**
   * Last good listings, from [`rememberListings`]. Omitted by callers that
   * report on one server at a time: the servers screen writes the failure
   * beside the row, and remembered sessions there would contradict it.
   */
  remembered: Readonly<Record<string, RememberedListing>> = {},
): CensusRow[] {
  const rows: CensusRow[] = [];
  for (const report of reports) {
    const outcome = report.outcome;
    const sessions =
      outcome.state === "listed"
        ? outcome.sessions
        : (remembered[report.server_id]?.sessions ?? []);
    for (const session of sessions) {
      rows.push({
        serverId: report.server_id,
        serverLabel: report.server_label,
        reachable: outcome.state === "listed",
        session,
      });
    }
  }
  // Unreachable rows sort with the ones nobody can attach to. They stay in the
  // list, so the eye still finds them inside their own project.
  const openable = (row: CensusRow): boolean => row.reachable && row.session.ready;
  return rows.sort((left, right) => {
    if (openable(left) !== openable(right)) return openable(left) ? -1 : 1;
    const byServer = left.serverLabel.localeCompare(right.serverLabel, "ko");
    if (byServer !== 0) return byServer;
    const byTitle = sessionTitle(left.session).localeCompare(sessionTitle(right.session), "ko");
    if (byTitle !== 0) return byTitle;
    return left.session.session_id.localeCompare(right.session.session_id, "ko");
  });
}

/** 대답하지 못한 서버들. 순서는 요청한 순서 그대로 — 새로고침마다 뒤섞이지 않게. */
export function failures(reports: readonly ServerReport[]): CensusFailure[] {
  return reports
    .map(describeFailure)
    .filter((failure): failure is CensusFailure => failure !== undefined);
}

/** 실제로 대답한 서버 수. */
export function answeredCount(reports: readonly ServerReport[]): number {
  return reports.filter((report) => report.outcome.state === "listed").length;
}

/**
 * 목록이 비었을 때 화면에 쓸 문장.
 *
 * "세션이 없습니다"와 "아무 서버도 대답하지 않았습니다"는 정반대의 뜻이고,
 * 전자를 후자의 상황에서 보여주면 사용자는 에이전트가 전부 죽었다고 읽는다.
 * 그래서 빈 화면 문구는 계산해서 고른다.
 */
export function emptyMessage(reports: readonly ServerReport[]): string {
  if (reports.length === 0) return "등록된 서버가 없습니다";
  if (answeredCount(reports) === 0) return "대답한 서버가 없습니다 — 아래 이유를 확인하세요";
  return "실행 중인 세션이 없습니다";
}

File: mobile/src-tauri/src/census.rs
//! Asking every server, directly, what sessions it is running.
//!
//! After pairing the laptop is gone. There is no index, no broker and nothing
//! to ask "where are my sessions" — the phone holds a list of servers and it
//! talks to each of them itself, over its own SSH connection, with its own
//! pinned host key. This module is the fan-out.
//!
//! ## The failure this module exists to prevent
//!
//! **A missing session reads as a dead session.** If one box is slow, off the
//! network, or has no `hmux` on its `PATH`, the naive shape — gather what came
//! back, show it — produces a shorter list with nothing to distinguish it from
//! a correct one. The user sees that the agent they left running last night is
//! not there and concludes it died. So:
//!
//! - every server in the list produces exactly one [`ServerReport`], always,
//!   including the ones that never answered and the ones that were never
//!   reached at all;
//! - a server that runs out of budget is [`ProbeOutcome::TimedOut`], not
//!   absent;
//! - a server the scheduler never got to before the overall deadline is
//!   [`ProbeOutcome::NotAttempted`], which is a different sentence from "that
//!   box did not answer" and belongs to a different fix.
//!
//! ## Not provisioned is not a protocol failure
//!
//! A box without `hmux` answers the exec channel with a shell's
//! `hmux: command not found` on stderr and closes. That is a complete, healthy
//! SSH session carrying a perfectly clear message, and reporting it as a
//! protocol or transport fault sends the user to look at the network. It is
//! classified as [`ProbeOutcome::NotProvisioned`], whose fix is on the server's
//! `PATH`.
//!
//! ## Why the scheduler is generic over the probe
//!
//! The interesting behaviour here is scheduling under two deadlines with
//! bounded parallelism, and none of it involves SSH. Taking the probe as a
//! closure is what makes "one wedged server does not remove the others" and
//! "every server is accounted for" testable at all — against a fake probe that
//! blocks, rather than against a lab that would have to be *made* to wedge.

use crate::catalog::DiscoveredSession;
use crate::relay::RelayError;
use serde::Serialize;
use std::collections::VecDeque;
use std::sync::Mutex;
use std::time::{Duration, Instant};

/// How many servers this client dials at once.
///
/// Four, matching the desktop's own census width (`MAX_CENSUS_WORKERS`). The
/// bound is about the device, not the servers: each worker holds an SSH session
/// with its own tokio runtime and thread, and a phone that opens twenty at once
/// on a cellular link makes every one of them slower.
pub const MAX_PARALLEL_SERVERS: usize = 4;

/// How long the whole census may take before the remaining servers are
/// reported as unattempted.
///
/// Deliberately not `MAX_PARALLEL_SERVERS × per-server budget`: with more
/// servers than workers the queue is serial, and a user watching a spinner
/// should get a partial answer with the stragglers *named* rather than a
/// complete one eventually.
pub const CENSUS_BUDGET: Duration = Duration::from_secs(25);

/// What happened when this device asked one server.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "snake_case", tag = "state")]
pub enum ProbeOutcome {
    Listed {
        sessions: Vec<DiscoveredSession>,
        /// 강제 명령이 실제로 적용됐는지, 게이트웨이가 관측해 보고한 값.
        ///
        /// 세션이 아니라 **연결**의 사실이라 여기 있다. `None`은 그 필드를
        /// 보내지 않는 예전 게이트웨이라 "모른다"는 뜻이고, 화면은 그것을
        /// `false`로 바꿔 말하지 않는다 — 멀쩡히 고정된 서버를 의심하게 만든다.
        forced_command_applied: Option<bool>,
    },
    /// SSH worked and `hmux` did not run. The fix is the server's `PATH` or its
    /// `authorized_keys` line, not its network.
    NotProvisioned { detail: String },
    /// This device cannot dial the server at all — no key stored, no pinned
    /// fingerprint. The fix is on the phone.
    NotConfigured { code: String, detail: String },
    /// The bytes did not get there, or the far end was not who it claimed.
    Unreachable { code: String, detail: String },
    /// The server answered the handshake and then did not finish.
    TimedOut { seconds: u64 },
    /// The census ran out of time before this server's turn came. Not the same
    /// sentence as "it did not answer", and not the same fix.
    NotAttempted,
}

impl ProbeOutcome {
    /// Sessions this outcome contributes to the merged list. Empty for every
    /// failure, so a caller cannot accidentally treat a failure as an empty
    /// server.
    #[must_use]
    pub fn sessions(&self) -> &[DiscoveredSession] {
        match self {
            Self::Listed { sessions, .. } => sessions,
            _ => &[],
        }
    }

    /// 강제 명령이 실제로 적용됐는지, 이 서버가 관측해 보고한 값.
    ///
    /// 대답하지 못한 서버는 `None`이다 — 물어보지 못한 것과 "적용되지 않았다"는
    /// 다르고, 후자로 말하면 닿지도 못한 서버를 안전하지 않다고 비난하게 된다.
    #[must_use]
    pub fn forced_command_applied(&self) -> Option<bool> {
        match self {
            Self::Listed {
                forced_command_applied,
                ..
            } => *forced_command_applied,
            _ => None,
        }
    }

    /// Whether this server actually answered. The UI needs this to decide
    /// between "no sessions anywhere" and "nothing answered", which are
    /// opposite messages.
    #[must_use]
    pub fn answered(&self) -> bool {
        matches!(self, Self::Listed { .. })
    }
}

/// Classifies a relay failure into what the user should do about it.
///
/// The mapping is the point of this function: `hmux` missing and the network
/// being down are the same "the channel closed" event at the transport layer,
/// and they have nothing in common at the layer a person acts on.
#[must_use]
pub fn classify(error: &RelayError) -> ProbeOutcome {
    match error {
        RelayError::Gateway { detail } => ProbeOutcome::NotProvisioned {
            detail: detail.clone(),
        },
        RelayError::Configuration { code, detail } => ProbeOutcome::NotConfigured {
            code: (*code).to_string(),
            detail: detail.clone(),
        },
        RelayError::TimedOut { after } => ProbeOutcome::TimedOut {
            seconds: after.as_secs(),
        },
        other => ProbeOutcome::Unreachable {
            code: other.code().to_string(),
            detail: other.to_string(),
        },
    }
}

/// A server to ask, named well enough to report on without a second lookup.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CensusTarget {
    pub server_id: String,
    pub server_label: String,
}

/// Exactly one of these per server in the list. Always.
#[derive(Clone, Debug, Serialize)]
pub struct ServerReport {
    pub server_id: String,
    pub server_label: String,
    pub outcome: ProbeOutcome,
}

/// Asks every target, at most `parallelism` at a time, inside `budget`.
///
/// Order of the returned reports matches the order of `targets`, so the UI
/// renders a stable list rather than one that reshuffles by whichever server
/// was quickest this time.
pub fn take_census<P>(
    targets: &[CensusTarget],
    parallelism: usize,
    budget: Duration,
    probe: P,
) -> Vec<ServerReport>
where
    P: Fn(&CensusTarget) -> ProbeOutcome + Sync,
{
    if targets.is_empty() {
        return Vec::new();
    }
    let deadline = Instant::now() + budget;
    let queue: Mutex<VecDeque<usize>> = Mutex::new((0..targets.len()).collect());
    let collected: Mutex<Vec<Option<ProbeOutcome>>> = Mutex::new(vec![None; targets.len()]);

    // `max(1)` because a parallelism of zero would drain nothing and report
    // every server as unattempted — a configuration mistake that would look
    // exactly like a total network outage.
    let workers = parallelism.clamp(1, targets.len());
    std::thread::scope(|scope| {
        for _ in 0..workers {
            scope.spawn(|| loop {
                let Some(index) = queue.lock().expect("census queue").pop_front() else {
                    return;
                };
                // Checked before starting, not after: a worker that begins a
                // 12-second dial one millisecond before the deadline would
                // hold the whole census open for those 12 seconds.
                let outcome = if Instant::now() >= deadline {
                    ProbeOutcome::NotAttempted
                } else {
                    probe(&targets[index])
                };
                collected.lock().expect("census results")[index] = Some(outcome);
            });
        }
    });

    let mut collected = collected.into_inner().expect("census results");
    targets
        .iter()
        .enumerate()
        .map(|(index, target)| ServerReport {
            server_id: target.server_id.clone(),
            server_label: target.server_label.clone(),
            // `NotAttempted` rather than an `expect`: every worker writes its
            // slot before returning, so this is unreachable today. It stays a
            // total function because the failure it would otherwise produce —
            // a panic on a phone — is far worse than one over-cautious row,
            // and because "a server vanished from the report" is precisely the
            // outcome this module exists to make impossible.
            outcome: collected[index]
                .take()
                .unwrap_or(ProbeOutcome::NotAttempted),
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::catalog::RemoteSession;
    use hmux_client::ClientError;
    use std::sync::atomic::{AtomicUsize, Ordering};

    fn targets(count: usize) -> Vec<CensusTarget> {
        (0..count)
            .map(|index| CensusTarget {
                server_id: format!("srv-{index}"),
                server_label: format!("서버 {index}"),
            })
            .collect()
    }

    fn session(id: &str) -> DiscoveredSession {
        RemoteSession {
            host_liveness: None,
            launch_program: None,
            session_id: id.to_string(),
            session_name: None,
            workspace_id: "ws-1".to_string(),
            session_class: Default::default(),
            lifecycle: Default::default(),
            provider_id: "claude".to_string(),
            runner_principal: "kattpish".to_string(),
            runner_instance: "run-1".to_string(),
            channel_epoch: "1".to_string(),
            host_instance_id: "host-1".to_string(),
            terminal_epoch: "term-1".to_string(),
            capabilities: Vec::new(),
        }
        .into()
    }

    #[test]
    fn every_target_produces_exactly_one_report_in_the_order_given() {
        let targets = targets(5);

        let reports = take_census(&targets, 2, Duration::from_secs(5), |target| {
            ProbeOutcome::Listed {
                forced_command_applied: None,
                sessions: vec![session(&target.server_id)],
            }
        });

        assert_eq!(reports.len(), 5);
        assert_eq!(
            reports
                .iter()
                .map(|report| report.server_id.as_str())
                .collect::<Vec<_>>(),
            vec!["srv-0", "srv-1", "srv-2", "srv-3", "srv-4"]
        );
    }

    /// The core requirement. One box that never answers must not remove the
    /// sessions on every other box — and must itself be named, not omitted.
    #[test]
    fn one_wedged_server_does_not_take_the_rest_of_the_list_with_it() {
        let targets = targets(3);

        let reports = take_census(&targets, 3, Duration::from_millis(150), |target| {
            if target.server_id == "srv-1" {
                std::thread::sleep(Duration::from_secs(3));
                return ProbeOutcome::TimedOut { seconds: 3 };
            }
            ProbeOutcome::Listed {
                forced_command_applied: None,
                sessions: vec![session(&target.server_id)],
            }
        });

        assert!(reports[0].outcome.answered());
        assert!(reports[2].outcome.answered());
        assert!(
            matches!(reports[1].outcome, ProbeOutcome::TimedOut { .. }),
            "the wedged server must be named: {:?}",
            reports[1].outcome
        );
    }

    /// A server the scheduler never reached is a different sentence from a
    /// server that did not answer, and the user acts on it differently.
    /// Reporting it at all is the whole point: silence would look like a box
    /// with no sessions.
    #[test]
    fn servers_the_budget_never_reached_are_named_rather_than_dropped() {
        let targets = targets(6);
        let attempts = AtomicUsize::new(0);

        let reports = take_census(&targets, 1, Duration::from_millis(120), |target| {
            attempts.fetch_add(1, Ordering::SeqCst);
            std::thread::sleep(Duration::from_millis(80));
            ProbeOutcome::Listed {
                forced_command_applied: None,
                sessions: vec![session(&target.server_id)],
            }
        });

        assert_eq!(reports.len(), 6, "every server must appear");
        assert!(
            attempts.load(Ordering::SeqCst) < 6,
            "the budget should have stopped the queue"
        );
        assert!(
            reports
                .iter()
                .any(|report| matches!(report.outcome, ProbeOutcome::NotAttempted)),
            "{reports:?}"
        );
    }

    /// The bound that keeps a phone from opening one SSH session per server at
    /// once. Without it, a list of twenty servers is twenty simultaneous
    /// handshakes on a cellular link.
    #[test]
    fn no_more_than_the_permitted_number_of_servers_are_dialed_at_once() {
        let targets = targets(8);
        let live = AtomicUsize::new(0);
        let peak = AtomicUsize::new(0);

        take_census(&targets, 3, Duration::from_secs(10), |target| {
            let now = live.fetch_add(1, Ordering::SeqCst) + 1;
            peak.fetch_max(now, Ordering::SeqCst);
            std::thread::sleep(Duration::from_millis(20));
            live.fetch_sub(1, Ordering::SeqCst);
            ProbeOutcome::Listed {
                forced_command_applied: None,
                sessions: vec![session(&target.server_id)],
            }
        });

        assert!(peak.load(Ordering::SeqCst) <= 3, "{peak:?}");
    }

    /// A parallelism of zero would drain nothing and report a total outage.
    #[test]
    fn a_parallelism_of_zero_still_asks_every_server() {
        let targets = targets(2);

        let reports = take_census(&targets, 0, Duration::from_secs(5), |_| {
            ProbeOutcome::Listed {
                forced_command_applied: None,
                sessions: Vec::new(),
            }
        });

        assert!(reports.iter().all(|report| report.outcome.answered()));
    }

    #[test]
    fn an_empty_server_list_is_an_empty_census_not_a_panic() {
        assert!(take_census(&[], 4, Duration::from_secs(1), |_| {
            ProbeOutcome::NotAttempted
        })
        .is_empty());
    }

    /// A box without `hmux` answers with a shell error over a perfectly healthy
    /// SSH session. Calling that a protocol failure sends the user to look at
    /// the network.
    #[test]
    fn a_server_without_hmux_is_not_provisioned_rather_than_a_protocol_failure() {
        let outcome = classify(&RelayError::Gateway {
            detail: "bash: hmux: command not found".to_string(),
        });

        assert!(
            matches!(outcome, ProbeOutcome::NotProvisioned { .. }),
            "{outcome:?}"
        );
    }

    #[test]
    fn a_missing_key_is_the_phones_problem_not_the_servers() {
        let outcome = classify(&RelayError::Configuration {
            code: "relay_identity_missing",
            detail: "이 서버에 등록된 SSH 개인키가 없습니다".to_string(),
        });

        match outcome {
            ProbeOutcome::NotConfigured { code, .. } => assert_eq!(code, "relay_identity_missing"),
            other => panic!("unexpected outcome: {other:?}"),
        }
    }

    #[test]
    fn a_session_level_refusal_is_reported_with_its_own_code() {
        let outcome = classify(&RelayError::Session(ClientError::MissingCapability {
            capability: "terminal_surface_v1",
        }));

        match outcome {
            ProbeOutcome::Unreachable { code, .. } => assert!(!code.is_empty()),
            other => panic!("unexpected outcome: {other:?}"),
        }
    }

    #[test]
    fn a_timeout_keeps_its_own_classification() {
        let outcome = classify(&RelayError::TimedOut {
            after: Duration::from_secs(12),
        });

        assert!(
            matches!(outcome, ProbeOutcome::TimedOut { seconds: 12 }),
            "{outcome:?}"
        );
    }

    /// A failed probe must contribute no sessions, so a caller merging the
    /// reports cannot turn a failure into "that server has nothing running".
    #[test]
    fn a_failed_probe_contributes_no_sessions_and_does_not_count_as_answered() {
        let outcome = ProbeOutcome::NotProvisioned {
            detail: "no hmux".to_string(),
        };

        assert!(outcome.sessions().is_empty());
        assert!(!outcome.answered());
    }
}
